### PR TITLE
feat: add agentic RAG with iterative sub-agent search

### PR DESCRIPTION
**Pull request by MattiaMolon**
## Summary
- Introduces a search sub-agent that iteratively queries the knowledge base to gather sufficient context before answering, replacing the single-shot retrieval approach
- Parallel tool calls are executed concurrently via `ThreadPoolExecutor`, with context proportionally allocated across results within the model's window
- Adds `async_rag` to mirror the sync `rag` function with full async streaming support

## Key changes
- **Sub-agent loop** (`_run_tool`): when `search_knowledge_base` is called, a dedicated agent with `SEARCH_AGENT_PROMPT` iterates up to `config.agentic_iterations`, deduplicating chunk spans by chunk ID across iterations
- **Token budget fix**: buffer now counts all messages (not just last per role) and reserves space for the LLM output, preventing context overflow
- **Robustness fixes**: safe fallback in `_clip` always preserves at least the last user message; `_get_tools` handles `None` content; `_limit_chunkspans` guards against zero-token edge case
- **Prompt improvements**: `SEARCH_AGENT_PROMPT` guides the sub-agent with concrete good/bad query examples; `NO_TOOLS_FOLLOW_UP_PROMPT` prevents tool calls in the final answer step

## Tests
- `test_rag_manual` — manual retrieval via `retrieve_context` + `add_context`, asserts no tool calls are made
- `test_rag_auto_with_retrieval` — agentic RAG on a question that requires retrieval; checks tool messages appear and `on_retrieval` callback is populated
- `test_rag_auto_without_retrieval` — agentic RAG on a trivial question; verifies no retrieval occurs
- `test_retrieve_context_self_query` — `retrieve_context` with `self_query=True`; asserts metadata filters are applied correctly
- `test_agentic_search_threads_metadata_filter_to_nested_tool_calls` — unit test verifying `metadata_filter` is forwarded from `search_knowledge_base` down into nested `query_knowledge_base` calls inside `_run_tool`
- `test_query_tool_call_passes_metadata_filter_to_retrieve_context` — unit test verifying `_run_tool` passes `metadata_filter` to `retrieve_context` for direct `query_knowledge_base` calls
- `test_sub_agent_deduplicates_chunk_spans_by_chunk_id` — unit test verifying that fully-redundant spans are dropped across sub-agent iterations while partially-novel spans are kept
- `test_rag_does_not_mutate_caller_messages_on_stream_error` — ensures the caller's `messages` list is unchanged when an exception is raised mid-stream

### CODE REVIEW

File: src/raglite/_rag.py [hunk 1]
```diff
@@ -285,15 +308,95 @@ def _get_tools(
...
+                model=config.llm,
+                messages=messages,
+                tools=[tool],
+                tool_choice="required" if iteration_index == 0 else "auto",
+            )
+            messages.append(response.choices[0].message.to_dict())  # type: ignore[arg-type,union-attr]
+            tool_calls = response.choices[0].message.tool_calls  # type: ignore[union-attr]
+
+            # check if the tool call is valid
+            if tool_calls is not None:
```
Comment (emilradix) [inline src/raglite/_rag.py]: an empty list ([]) will pass this check i believe. Is that intentional?

File: src/raglite/_rag.py [hunk 2]
```diff
@@ -285,15 +308,95 @@ def _get_tools(
...
+            },
+        }
+
+        # Start iterating and keep only chunk spans that introduce at least one new chunk ID.
+        chunk_spans: list[ChunkSpan] = []
+        seen_chunk_ids: set[str] = set()
+        for iteration_index in range(max(1, config.agentic_iterations)):
+            response = completion(
+                model=config.llm,
+                messages=messages,
```
Comment (emilradix) [inline src/raglite/_rag.py]: should these messages have _clip appleid?

**Review by emilradix — COMMENTED** [review]
A few comments. Can you double check the github copilot comments as well.

File: src/raglite/_rag.py [hunk 3]
```diff
@@ -285,15 +308,95 @@ def _get_tools(
...
+                retrieved_chunk_spans: list[ChunkSpan] = []
+                messages.extend(
+                    _run_tools(
+                        tool_calls,
+                        retrieved_chunk_spans.extend,
+                        config,
+                        messages=messages,
+                        metadata_filter=metadata_filter,
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (MattiaMolon) [reply src/raglite/_rag.py:386]: The nesting is only one level deep and the workload is small:
- Outer executor: typically 1-2 search_knowledge_base calls
- Inner executor (per search call): up to ~3 query_knowledge_base calls per iteration
So worst case you'd have ~6 concurrent threads total, within ThreadPoolExecutor's default limit (min(32, cpu_count + 4)). The executors also don't share thread pools, each manages its own, so there should be no deadlock risk either.

File: tests/test_rag.py [hunk 4]
```diff
@@ -78,3 +88,195 @@ def test_retrieve_context_self_query(raglite_test_config: RAGLiteConfig) -> None
...
+    monkeypatch: pytest.MonkeyPatch,
+) -> None:
+    """Pass metadata filters from search_knowledge_base down to nested tool calls."""
+    config = RAGLiteConfig(
+        llm="gpt-5-mini",
+        embedder="text-embedding-3-small",
+        db_url="duckdb:///:memory:",
+    )
+    metadata_filter: MetadataFilter = {"topic": ["Physics"]}
+    nested_tool_call = SimpleNamespace(
```
[the remark this answers is not in the document]
Comment (MattiaMolon) [reply tests/test_rag.py:100]: changed in new commit

File: tests/test_rag.py [hunk 5]
```diff
@@ -40,9 +47,12 @@ def test_rag_auto_with_retrieval(raglite_test_config: RAGLiteConfig) -> None:
...
     assert "event" in answer.lower()
     # Verify that RAG context was retrieved automatically.
-    assert [message["role"] for message in messages] == ["user", "assistant", "tool", "assistant"]
-    assert json.loads(messages[-2]["content"])
+    roles = [message["role"] for message in messages]
+    assert roles[0] == "user"
+    assert roles[-1] == "assistant"
+    assert "tool" in roles  # At least one retrieval happened.
     if not raglite_test_config.llm.startswith("llama-cpp-python"):
+        assert "tool" in roles  # At least one retrieval happened.
```
[the remark this answers is not in the document]
Comment (MattiaMolon) [reply tests/test_rag.py]: removed assertion in new commit

File: src/raglite/_rag.py (hunk 1 above)
Comment (MattiaMolon) [reply src/raglite/_rag.py]: solved in new commit

File: src/raglite/_rag.py (hunk 2 above)
Comment (MattiaMolon) [reply src/raglite/_rag.py]: added in new commit

File: src/raglite/_rag.py [hunk 6]
```diff
@@ -285,15 +308,95 @@ def _get_tools(
...
+    *,
+    metadata_filter: MetadataFilter | None = None,
 ) -> tuple[str, list[ChunkSpan]]:
     """
     Run a single tool to search the knowledge base.
 
     Returns the tool_id and the raw chunk_spans (before formatting/limiting).
     """
     if tool_call.function.name == "search_knowledge_base":
+        query = json.loads(tool_call.function.arguments)["query"]
```
[the remark this answers is not in the document]
Comment (MattiaMolon) [reply src/raglite/_rag.py]: added and simplified in new commit

**Review by emilradix — APPROVED** [review]
LGTM
